### DISCUSSION TITLE
Idea of leveraging CUDA for computing

### DISCUSSION THREAD

**Discussion by ray1422** [bevyengine/bevy#16232] [Ideas]
Hi dear team,

I am a big fan of the game Cities Skylines, which is an agent-based city simulation game and I'm wondering if I can leverage CUDA to improve the performance by parallel computing.

The rough idea is to maintain some data, or more accurately, components, in CUDA, and write CUDA kernel as Systems.

I wrote a PoC of lock-based car driving on path simulation with CUDA, which is significantly faster than computing on a CPU. I define paths, and each path owns candidates. Every car purpose they want to drive on Path(start, end, priority) and utilize CUDA threads to decide which cars acquire the path interval.
If I can make values on GPU memory _flow_ to render part (interop) without copying between host and device (CPU & GPU), this approach will significantly improve the performance. Based on my calculation, it can simulate over 1000 km road network, with 10-100M Cims, which is as large as some city like Singapore of Taipei, making the game more realistic.

Are there any idea to leverage CUDA computing with Bevy engine?

Thanks

**Comment by Touma-Kazusa2** [comment]
If you want to use the GPU for computation, maybe you can use wgpu(bevy use it) compute pass?I'm not particularly knowledgeable in this area.
